### PR TITLE
feat(atomic): generic Search/Facet Search Component

### PR DESCRIPTION
**Pull request by nrawji**
https://coveord.atlassian.net/browse/KIT-431

Apologies for the biggish PR. This PR refactors facet search & regular search box into one `base-search` component. I also had to modify headless to add a clearResults method on Facet Search in order to hide results on FacetSearch blur (this is how the search box hides suggestions). I could also just handle this internally inside atomic (the downside of this is that it's slightly more complicated and asymmetrical with the regular search box.

### CODE REVIEW

File: packages/atomic/src/utils/combobox.tsx [hunk 1]
```diff
@@ -161,9 +161,10 @@ export class Combobox {
 
     this.removeEmptyOptionElement();
 
-    Array.from(this.listboxOptions).forEach((value) =>
-      this.updateOption(value)
-    );
+    Array.from(this.listboxOptions).forEach((value, index) => {
+      value.setAttribute('id', `${this.options.id}-suggestion-${index}`);
```
Comment (btaillon) [inline packages/atomic/src/utils/combobox.tsx:165]: It's a small preference, but I would prefer if this was moved inside `updateOption` and an `index`/`id` parameter was added to `updateOption`.

**Review by btaillon — APPROVED** [review]
Looks good

File: packages/atomic/src/components/facets/facet-search/facet-search.pcss [hunk 2]
```diff
@@ -0,0 +1,46 @@
+.search-box {
+  @apply relative;
+}
+
+.search-box-wrapper, .suggestion {
+  height: var(--atomic-facet-search-height, 30px);
+}
+
+.search-box-wrapper {
+  @apply apply-border-on-background rounded  focus-within:rounded-b-none;
```
[the remark this answers is not in the document]
Comment (nrawji) [reply packages/atomic/src/components/facets/facet-search/facet-search.pcss:10]: I guess just to explain the way the `base-search-box` component is supposed to work, its a generic search box with the proper accessibility configuration that can be used basically anywhere we need a search box. I tried to keep the styles it contains relatively limited only containing classes that help position elements.

The intention is that other components can contain the `base-search-box` component and then style is however they want by providing their own classes for it. So facet search might have in its CSS something like
```
.search-box-wrapper {
  @apply bg-background mx-2;
  height: 30px;
}
```
and the main search box could provide its own CSS for the same `.search-box-wrapper` allowing facet search and the search box to be styled differently but maintaining the same HTML, accessibility stuf and general logic.

File: packages/atomic/src/components/searchbox/base-search.tsx [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+import {I18nState} from '../../utils/initialization-utils';
+import {Combobox} from '../../utils/combobox';
+
+import ClearIcon from 'coveo-styleguide/resources/icons/svg/clear.svg';
+import SearchIcon from 'coveo-styleguide/resources/icons/svg/search.svg';
+
+@Component({
+  tag: 'base-search',
+  styleUrl: 'base-search.pcss',
+  shadow: false,
```
[the remark this answers is not in the document]
Comment (nrawji) [reply packages/atomic/src/components/searchbox/base-search.tsx:11]: Since it's an internal component, it will always be encapsulated into the shadow dom of the component that contains it. We don't really want or need nested shadow doms (this will just make it harder for us to style internal components).

File: packages/atomic/src/components/searchbox/base-search.tsx [hunk 4]
```diff
@@ -0,0 +1,182 @@
+import {Component, Event, EventEmitter, h, Prop} from '@stencil/core';
+import {I18nState} from '../../utils/initialization-utils';
+import {Combobox} from '../../utils/combobox';
+
+import ClearIcon from 'coveo-styleguide/resources/icons/svg/clear.svg';
+import SearchIcon from 'coveo-styleguide/resources/icons/svg/search.svg';
+
+@Component({
+  tag: 'base-search',
```
[the remark this answers is not in the document]
Comment (nrawji) [reply packages/atomic/src/components/searchbox/base-search.tsx:9]: Yeah just to reiterate what I stated in stand-up, this can probably be replaced with a functional component but it'll take some refactoring in how we handle the combobox class. So for now we're going to keep it as an internal full fledged web component, in order to start writing tests and let Lucille/William start testing it.

File: packages/headless/src/features/facets/facet-search-set/facet-search-reducer-helpers.ts [hunk 5]
```diff
@@ -54,6 +54,19 @@ export function handleFacetSearchUpdate<T extends FacetSearchResponse>(
...
+  state: FacetSearchSetState<T>,
+  facetId: string
+) {
+  const search = state[facetId];
+
+  if (!search) {
+    return;
+  }
+
+  search.response.values = [];
```
[the remark this answers is not in the document]
Comment (nrawji) [reply packages/headless/src/features/facets/facet-search-set/facet-search-reducer-helpers.ts:67]: This is how the search box suggestions were implemented in Headless (there are `showSuggestions` and `hideSuggestions` in the Headless Search Box which just clear the values array), I decided to keep the behaviour consistent, but I generally agree that this is more of a UI problem than anything.

[the remark this answers is not in the document]
Comment (samisayegh) [reply packages/headless/src/features/facets/facet-search-set/facet-search-reducer-helpers.ts:67]: Maybe this is something we can improve with the headless search box controller too? It might need to wait for 1.x though if it is breaking change.

In general, if an existing pattern leads to a slower experience, or unnecessary queries, I would prefer to avoid the pattern rather than continue down the path.

[the remark this answers is not in the document]
Comment (nrawji) [reply packages/headless/src/features/facets/facet-search-set/facet-search-reducer-helpers.ts:67]: I mean we could just mark the methods as deprecated inside headless and not use them in atomic.

File: packages/atomic/src/components/atomic-search-box/atomic-search-box.pcss [hunk 6]
```diff
@@ -1,6 +1,10 @@
 @import "tailwindcss/base";
 @import "tailwindcss/utilities";
 
+atomic-search-box {
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/atomic-search-box/atomic-search-box.pcss:4]: why is this necessary? Could we keep a tailwind class `z-50` on the suggestions?

File: packages/atomic/src/components/searchbox/base-search.tsx [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+      <button
+        class="px-1 text-primary"
+        onClick={() => this.showMoreResults.emit()}
+      >
+        show more
+      </button>
+    );
+  }
+
+  public render() {
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/searchbox/base-search.tsx:151]: I would split this in multiple render methods for visibility

**Comment by ThibodeauJF** [comment]
Strings should be localized, like for the "show more"

**Review by samisayegh — APPROVED** [review]

**Comment by ThibodeauJF** [comment]
I'm not 100% sure of this change, the `base-search` component kind of makes the combobox helper class pointless and improves complexity, here we now have:

- search-box -> base-search -> combobox
- facets (all type) -> facet-search -> base-search -> combobox

I think 2 levels should be sufficient, it's much easier to grasp and maintain, either do: 
1. search-box/facet-search + combobox
2. merge base-search & combobox

Personally I would go with 1, I could argue that there are sufficient differences with those 2 searches that could justify not using a common base-search component, such as the show more/less, a submit button, possibly having a show more/less on scroll in the facet search or displaying previews in suggestions, etc. When we will have to add more complex features to suggestions or facet search results, we're going to have more and more complexity to the base-search component.

Side-note: Do the css parts still work correctly? They might because shadow DOM is off, but they might also need to be exposed.
